### ISSUE TITLE
Native HubPro Support

### ISSUE THREAD

**Issue by rasod** [kuestess/homebridge-platform-insteonlocal#13]
Less of an issue with the plugin, more of an "[nerd snipe](https://xkcd.com/356/)" question.

I have been trying to get the plugin to run natively on the HubPro, since I got a pile of them (if you count 3 a pile) and if it worked it would allow instant status updates. The HubPro is basically a Beagle Board with a PLM hooked up via serial pins. I managed to boot the BB off an SD and install the plugin. However can't seem to get the serial connection to return data home-controller can understand. Not sure if it is a baud/bit rate or just not possible.

```javascript
  home-controller:insteon connecting /dev/ttyS4 +0ms
  home-controller:insteon connected +204ms
Connected!
Init
  home-controller:insteon Send: {"command":{"raw":"0260","type":"60"},"deferred":{"promise":{}},"timeout":5000} +95ms
  home-controller:insteon insteon.links - id: null +39ms
  home-controller:insteon Already running a command - queuing command: {"raw":"0269","type":"69"} +6ms
  home-controller:insteon Rcvd: ff +100ms
  home-controller:insteon checkStatus - buffer: ff +6ms
  home-controller:insteon checkStatus - status: {"command":{"raw":"0260","type":"60"},"deferred":{"promise":{}},"timeout":5000} +3ms
  home-controller:insteon parsing - insufficient data to continue - 2 +3ms
  home-controller:insteon sendCommandTimout - status: {"command":{"raw":"0260","type":"60"},"deferred":{"promise":{}},"timeout":5000} +5s
  home-controller:insteon Resend: {"command":{"raw":"0260","type":"60"},"deferred":{"promise":{}},"timeout":5000} +4ms
  home-controller:insteon Rcvd: ff +20ms
  home-controller:insteon checkStatus - buffer: ffff +3ms
  home-controller:insteon checkStatus - status: {"command":{"raw":"0260","type":"60"},"deferred":{"promise":{}},"timeout":5000} +3ms
  home-controller:insteon parsing - insufficient data to continue - 4 +2ms
  home-controller:insteon sendCommandTimout - status: {"command":{"raw":"0260","type":"60"},"deferred":{"promise":{}},"timeout":5000} +5s
  home-controller:insteon Send: {"command":{"raw":"0269","type":"69"},"deferred":{"promise":{}},"timeout":5000} +14ms
Hub Info: undefined
  home-controller:insteon Rcvd: ff +36ms
  home-controller:insteon checkStatus - buffer: ffffff +4ms
  home-controller:insteon checkStatus - status: {"command":{"raw":"0269","type":"69"},"deferred":{"promise":{}},"timeout":5000} +2ms
  home-controller:insteon parsing - message skipped +9ms
  home-controller:insteon checkStatus - buffer: ff +6ms
  home-controller:insteon checkStatus - status: {"command":{"raw":"0269","type":"69"},"deferred":{"promise":{}},"timeout":5000} +2ms
  home-controller:insteon parsing - insufficient data to continue - 2 +3ms
  home-controller:insteon sendCommandTimout - status: {"command":{"raw":"0269","type":"69"},"deferred":{"promise":{}},"timeout":5000} +5s
  home-controller:insteon Resend: {"command":{"raw":"0269","type":"69"},"deferred":{"promise":{}},"timeout":5000} +23ms
  home-controller:insteon Rcvd: ff +26ms
  home-controller:insteon checkStatus - buffer: ffff +2ms
  home-controller:insteon checkStatus - status: {"command":{"raw":"0269","type":"69"},"deferred":{"promise":{}},"timeout":5000} +2ms
  home-controller:insteon parsing - insufficient data to continue - 4 +2ms
  home-controller:insteon sendCommandTimout - status: {"command":{"raw":"0269","type":"69"},"deferred":{"promise":{}},"timeout":5000} +5s
Links: 
  home-controller:insteon Rcvd: d9fc +22s
  home-controller:insteon checkStatus - buffer: ffffd9fc +2ms
  home-controller:insteon checkStatus - status: undefined +2ms
  home-controller:insteon parsing - message skipped +4ms
  home-controller:insteon checkStatus - buffer: d9fc +5ms
  home-controller:insteon checkStatus - status: undefined +3ms
  home-controller:insteon parsing - insufficient data to continue - 4 +2ms
  home-controller:insteon Rcvd: d9f9 +265ms
  home-controller:insteon checkStatus - buffer: d9fcd9f9 +2ms
  home-controller:insteon checkStatus - status: undefined +3ms
  home-controller:insteon parsing - message skipped +2ms
  home-controller:insteon checkStatus - buffer: d9f9 +5ms
  home-controller:insteon checkStatus - status: undefined +17ms
  home-controller:insteon parsing - insufficient data to continue - 4 +2ms
  home-controller:insteon Rcvd: 0849123857 +19s
  home-controller:insteon checkStatus - buffer: d9f90849123857 +3ms
  home-controller:insteon checkStatus - status: undefined +2ms
  home-controller:insteon parsing - message skipped +3ms
  home-controller:insteon checkStatus - buffer: 0849123857 +4ms
  home-controller:insteon checkStatus - status: undefined +2ms
  home-controller:insteon parsing - message skipped +3ms
  home-controller:insteon checkStatus - buffer: 123857 +3ms
  home-controller:insteon checkStatus - status: undefined +2ms
  home-controller:insteon parsing - message skipped +2ms
  home-controller:insteon checkStatus - buffer: 57 +3ms
  home-controller:insteon checkStatus - status: undefined +2ms
  home-controller:insteon parsing - insufficient data to continue - 2 +2ms
  home-controller:insteon Rcvd: 08 +206ms
  home-controller:insteon checkStatus - buffer: 5708 +2ms
  home-controller:insteon checkStatus - status: undefined +2ms
  home-controller:insteon parsing - insufficient data to continue - 4 +2ms
  home-controller:insteon Rcvd: f77d088c +4ms
  home-controller:insteon checkStatus - buffer: 5708f77d088c +2ms
  home-controller:insteon checkStatus - status: undefined +3ms
  home-controller:insteon parsing - message skipped +2ms
  home-controller:insteon checkStatus - buffer: f77d088c +4ms
  home-controller:insteon checkStatus - status: undefined +2ms
  home-controller:insteon parsing - message skipped +3ms
  home-controller:insteon checkStatus - buffer: 088c +3ms
  home-controller:insteon checkStatus - status: undefined +2ms
  home-controller:insteon parsing - insufficient data to continue - 4 +2ms
  home-controller:insteon Rcvd: 08 +4s
  home-controller:insteon checkStatus - buffer: 088c08 +2ms
  home-controller:insteon checkStatus - status: undefined +2ms
  home-controller:insteon parsing - message skipped +2ms
  home-controller:insteon checkStatus - buffer: 08 +4ms
  home-controller:insteon checkStatus - status: undefined +3ms
  home-controller:insteon parsing - insufficient data to continue - 2 +2ms
  home-controller:insteon Rcvd: 49323b47ff +3ms
  home-controller:insteon checkStatus - buffer: 0849323b47ff +2ms
  home-controller:insteon checkStatus - status: undefined +2ms
  home-controller:insteon parsing - message skipped +2ms
  home-controller:insteon checkStatus - buffer: 323b47ff +8ms
  home-controller:insteon checkStatus - status: undefined +2ms
  home-controller:insteon parsing - message skipped +3ms
  home-controller:insteon checkStatus - buffer: 47ff +3ms
  home-controller:insteon checkStatus - status: undefined +2ms
  home-controller:insteon parsing - insufficient data to continue - 4 +2ms
  home-controller:insteon Rcvd: 280449 +378ms
  home-controller:insteon checkStatus - buffer: 47ff280449 +3ms
  home-controller:insteon checkStatus - status: undefined +2ms
  home-controller:insteon parsing - message skipped +2ms
  home-controller:insteon checkStatus - buffer: 280449 +4ms
  home-controller:insteon checkStatus - status: undefined +2ms
  home-controller:insteon parsing - message skipped +3ms
  home-controller:insteon Rcvd: 1dfbfd +3ms
  home-controller:insteon checkStatus - buffer: 491dfbfd +2ms
  home-controller:insteon checkStatus - status: undefined +2ms
  home-controller:insteon parsing - message skipped +2ms
  home-controller:insteon checkStatus - buffer: fbfd +4ms
  home-controller:insteon checkStatus - status: undefined +3ms
  home-controller:insteon parsing - insufficient data to continue - 4 +2ms
  home-controller:insteon checkStatus - buffer: fbfd +2ms
  home-controller:insteon checkStatus - status: undefined +2ms
  home-controller:insteon parsing - insufficient data to continue - 4 +2ms
```

**Comment by kuestess** [comment]
@rasod home-controller has built-in serial support (which I haven’t implemented in the plugin).  I think it’s intended for a serial PLM but maybe that will work here. Try 19.2/8/N/1.

**Comment by rasod** [comment]
Bingo! 19200 is the magic number - I tried every baud I could think of, but 19200 since it is suppose to be the default for home-controller

**Comment by rasod** [comment]
It Works!

I changed the connection in the plugin to use the serial port and it works. Had to modify the connectionWatcher slightly - to prevent multiple ports being open.

It is fast changes are instant, even with the event listener disabled.

I have only tested it with one device so far. Still have to figure out how to link all the devices without having to go around the house pushing the set button. home-controller can setup the links I think have to try out how.

It's pretty easy.

1. Pop open the HubPro
2. Insert a microSD card that is Flashed with the BB software
3. Boot Up holding the S2 button down
4. SSH in
5. Enable the Serial Port and reboot
6. Install HomeBrige and Plugin as usual

I'll fork and do a PR if you are open to that.

**Comment by kuestess** [comment]
@rasod Wow, great work!  Happy to accept a PR to add support. I’m away right now, otherwise I’d drop everything and work on this. :-)

**Comment by rasod** [comment]
I had to write a little script to link the devices to the hub. If they are not linked you can control them and pull their status but the status won't be pushed as an event. We can probably make the plugin automatically create the links for new devices by comparing the config.json with the links object from the hub, but one thing at a time.

```javascript
var Insteon = require('home-controller').Insteon
var hub = new Insteon();

hub.serial('/dev/ttyS4',{baudRate:19200}, function(){
    console.log('Connected!');
    init();
});

function init(){
    console.log("Creating Links -- Here Comes The Beeps");
    hub.link(['30.69.FF','30.68.EE'], function(error, link) {})
}
```

**Comment by kuestess** [comment]
@rasod I wrote the `getDevices()` function with the intent of discovering devices automatically which should populate an array of devices that the Hub knows about.  I haven't touched it in awhile, but that should at least be a start.

**Comment by shereefalkady** [comment]
@rasod thank you so much -- it would have been a workout to beep each one manually TWICE....
will keep you updated.... just got my sd card will try the mission now

**Comment by shereefalkady** [comment]
@rasod again they beeped ... but now i am just staring at a line that says 

"Creating Links -- Here Comes The Beeps"

what am i supposed to do now

**Comment by rasod** [comment]
Going to close this issue because I think we have gotten it to work pretty well.

**Comment by shereefalkady** [comment]
hello @rasod im trying out insteonlocal again -- im unable to find how to set it up on the 2243 without needing the s2 button on every boot.  i know there was a post about it somewhere but i cant find it... any chance you can point me to it ?

**Comment by rasod** [comment]
@shereefalkady it has been a while - take horrible notes. I think I followed these instructions.

https://stackoverflow.com/questions/31725206/unable-to-flash-emmc-from-sd-card-beaglebone-black

**Comment by kuestess** [comment]
@rasod @shereefalkady This is what I followed as well.  Works like a charm.  You don't get the lights flashing like it indicates (from memory), but just wait ~10min to make sure its done.

**Comment by shereefalkady** [comment]
@rasod @kuestess is it possible i bricked my 2243?

**Comment by shereefalkady** [comment]
@rasod @kuestess ok update - it actually flashed the image onto the board... but that image is not good --- how do i flash another on the same board

**Comment by kuestess** [comment]
@shereefalkady You should just be able to put a new image onto the card and re-flash like you just  did.

**Comment by shereefalkady** [comment]
ok so all is working great ... i love all the improvements.... however ... I cant seem to get it to work right from systemd.... something with permissions - when i launch via command "homeridge -D" all works great -- when i launch from systemd gives me error: permission denied cannot open /dev/ttyS4

**Comment by shereefalkady** [comment]
[Unit]
Description=Homebridge
After=syslog.target network-online.target

[Service]
Type=simple
User=homebridge
EnvironmentFile=/etc/default/homebridge
ExecStart=$(which homebridge) \$HOMEBRIDGE_OPTS
Restart=on-failure
RestartSec=3
KillMode=process
CapabilityBoundingSet=CAP_IPC_LOCK CAP_NET_ADMIN CAP_NET_BIND_SERVICE CAP_NET_RAW CAP_SETGID CAP_SETUID CAP_SYS_CHROOT CAP_CHOWN CAP_FOWNER CAP_DAC_OVERRIDE CAP_AUDIT_WRITE CAP_SYS_ADMIN
AmbientCapabilities=CAP_NET_RAW

[Install]
WantedBy=multi-user.target
EOL



anything wrong there

**Comment by shereefalkady** [comment]
https://github.com/nfarina/homebridge/wiki/Install-Homebridge-on-Debian-or-Ubuntu-Linux

those were the isntructions followed

**Comment by shereefalkady** [comment]
@kuestess great job!! It’s come a long way ... 
I love the any % from zero and I love the fact that it doesn’t crash when a device is undisoverable now ... but I only tried with a 10house system - will try a bigger one soon

**Comment by kuestess** [comment]
@shereefalkady Welcome back - glad its working well for you.  Check out the UI as well.  Just load up `http://<homebridge IP>:<insteonlocal port>` in your browser.

**Comment by 1004PTLN** [comment]
@kuestess and @rasod first off I’m sorry if posting to a closed comment isn’t correct protocol but I’m new to GitHub and this comment seemed closest to my question.  Please advise if I should be posting differently.  Second, sorry for the length of the post, but wanted to be as clear as possible.   Lastly, thank you for your work on this project.   I have read through the readme and through all of the open and closed comments.  Clearly you have accomplished an amazing amount on this project.   Congratulations!

My question relates to the hub pro.  I’ve got a version 5 hub pro and after seeing the success you have had with flashing the beaglebone black I thought I’d give it a try.   The hub pro I have is brand new...in so far as I never actually put it in use. When I bought it, I was using ISY but really wanted to get on HomeKit as my family had embraced the Apple ecosystem.   When I saw all of the negative user feedback on the hub pro, I opted to shelve it and went into the Indigo system using their homebridge plug in, which is my current set up. Everything works fine, but to be honest all of my automations are in homekit not in Indigo, and I really don’t need to use the power of indigo itself so it seems like an unneeded annual expense to effectively use it only for their homebridge plugin.   Which brings me to my question.  

I have followed all of your steps to flash the customized hub pro beagle board but don’t seem to be making progress.  I have been able to create an microSD from the beagle bone site using balena etcher.  I have tried this with the newest flasher image on the site but saw another comment that said to try version 9.9 which I did as well.  That version was not available in a flasher version but frankly that didn’t matter because using both methods I was never able to seemingly complete the process and was not able to SSH into the system.  

The behavior of the board after placing the microSD in the slot, depressing S2 (boot button) and then applying power was that LED D1 (near the Ethernet port) shows constant green and LED D2 (near the microSD) shows a heartbeat pattern.  Seems like the standard board has a variety of other blue LEDs so that documentation doesn’t help. 

 I left both attempts run for over an hour after which I connected a USB data cable between the hub pro and a PC and used Putty to try to SSH into the hub pro and it simply times out.  

Can you provide any guidance here?   Am I doing something wrong that is obvious?   Does the LED pattern seem to make sense?   From the other standard board documentation it seems that once the flasher version is done the board should power down.  That didn’t happen when using the flasher version, but again I don’t know if that’s typical behavior of this custom board.   When using the non flasher version, eg the standard bootable version 9.9, LED 2 had the same behavior, eg consistent heartbeat pattern.  

I’m hoping it’s something simple I’m missing?   I’m using a data/transfer cable (not a power cable) to connect to the hub pro...which I assume is correct?  I don’t have a reason to use putty often so don’t know if it could be that simple that I’m using an incorrect cable?

Any guidance would be greatly appreciated.

**Comment by kuestess** [comment]
@1004PTLN I was going to suggest connecting via the network, and appears that worked for you.  Feel free to create an issue if you have other problems.
